js_src/_dart_helpers.js

```diff
@@ -83,10 +83,20 @@ function _createReactDartComponentClass2(dartInteropStatics, componentStatics, j
   class ReactDartComponent2 extends React.Component {
     constructor(props, context) {
       super(props, context);
-      // TODO combine these two calls into one
+      // TODO combine these two calls into one 
+      this.state = null;
       this.dartComponent = dartInteropStatics.initComponent(this, componentStatics);
-      this.state = dartInteropStatics.handleGetInitialState(this.dartComponent);
+      const initialStateFromGetInitialState =
+      dartInteropStatics.handleGetInitialState(this.dartComponent);
+      // If the returned state is empty, assume it's coming from the default implementation.
+      // This is done to avoid changing the default getInitialState implementation to null, which would cause breakages.
+      if (this.state && Object.keys(initialStateFromGetInitialState).length != 0) {
+        throw 'Error: State cannot be initialized in both init and getInitialState';
+      } else {
```

Comment:
RE my comment here https://github.com/cleandart/react-dart/pull/176#discussion_r276860480 

I think we can fix that by doing:
```suggestion
      } else if (!this.state) {
```